src/Workspaces/Core/Portable/Progress/CodeAnalysisProgress.cs

```diff
@@ -0,0 +1,147 @@
+
+    /// <summary>
+    /// When passed to an appropriate <see cref="IProgress{T}"/>, will add the requested number of incomplete items to
+    /// the UI showing the progress of the current operation.  This is commonly presented with a progress bar.  An
+    /// optional <paramref name="description"/> can also be provided to update the UI accordingly (see <see
+    /// cref="Description"/>).
+    /// </summary>
+    /// <param name="count">The number of incomplete items left to perform.</param>
+    /// <param name="description">Optional description to update the UI to.</param>
+    /// <example>
+    /// progress.Report(CodeAnalysisProgress.AddIncompleteItems(20));
+    /// </example>
+    public static CodeAnalysisProgress AddIncompleteItems(int count, string? description = null)
+        => new()
+        {
+            IncompleteItemsValue = count >= 0 ? count : throw new ArgumentOutOfRangeException(nameof(count)),
+            DescriptionValue = description,
+        };
+
+    /// <summary>
+    /// When passed to an appropriate <see cref="IProgress{T}"/>, will indicate that some items of work have
+    /// transitioned from being incomplete (see <see cref="AddIncompleteItems"/> to complete.  This is commonly
+    /// presented with a progress bar. An optional <paramref name="description"/> can also be provided to update the UI
+    /// accordingly (see <see cref="Description"/>).
+    /// </summary>
+    /// <param name="count">The number of items that were completed. Must be greater than or equal to 1.</param>
+    /// <param name="description">Optional description to update the UI to.</param>
+    /// <example>
+    /// progress.Report(CodeAnalysisProgress.CompleteItem());
+    /// </example>
+    public static CodeAnalysisProgress CompleteItems(int count, string? description = null)
+        => new()
+        {
+            CompleteItemValue = count >= 1 ? count : throw new ArgumentOutOfRangeException(nameof(count)),
+            DescriptionValue = description,
+        };
+
+    /// <summary>
+    /// When passed to an appropriate <see cref="IProgress{T}"/>, will indicate that all progress should be reset for
+    /// the current operation. This is normally done when the code action is performing some new phase and wishes for
+    /// the UI progress bar to restart from the beginning.
+    /// </summary>
+    /// <remarks>
+    /// Currently internal as only roslyn needs this in the impl of our suggested action (we use a progress bar to
+    /// compute the work, then reset the progress to apply all the changes).  Could be exposed later to 3rd party code
+    /// if a demonstrable need is presented.
+    /// </remarks>
+    internal static CodeAnalysisProgress Clear()
+        => new() { ClearValue = true };
+}
+
+internal sealed class CodeAnalysisProgressTracker(Action<string?, int, int>? updateAction) : IProgress<CodeAnalysisProgress>
```

Comment:
>CodeAnalysisProgressTracker [](http://example.com/codeflow?start=22&length=27)

Naming seems a bit odd here, seems like if it derives from IProgress<T>, it's name should be CodeAnalysisProgress (and the current class with that name should be something else)

Reply:
Renamed to CodeAnalysisProgressCallback

Reply:
Actually, just simplifie dthis up the whazhoo.